README
Libhdate, hcal, hdate
=====================

CONTENTS
========
1] Description
2] Brief highlights of this update
3] Installation from source (Instructions)


1] Description
==============
LibHdate is a small library for the Hebrew calendar, dates, holidays,
and reading sequence. It is written in C and includes bindings for
pascal, perl, python, php, ruby.

hcal and hdate are small example command line programs written in C.

Refer to the following files:
INSTALL      instructions how to install this package
USE          basic example of how to use one of the library functions
             in a C program, how to compile it, and how to run it.
NEWS         press-release style information of the current release
Changelog    more detailed list of project activity over many releases

Refer to the man pages and examples directory for further documentation
and examples.


2] Brief highlights of this update
==================================
This release brings many new options, features, and bug fixes to the two example
programs hcal and hdate. The changes to the underlying function library include
a few minor bug fixes, deprecation of a series of string functions in favor
of a single new one with better memory allocation, and hard-coding of core
elements of the Hebrew localization so that Hebrew can be displayed in all
locales. The documentation in the man pages and in the new configuration
file should be sufficient.

2.1] selected highlights
========================

hcal
- introduced config file $XDG_CONFIG_HOME/hcal/hcalrc or ~/.config/hcal/hcalrc
- option --menu -m to prompt user for user-defined menu selections from config file
- option --candles and --havdalah take optional numeric arguments to modify the default minhag.
- remove dependency of Hebrew on locale environment variables
- new feature: sunset-aware to highlight correct hebrew date
- new options: shabbat times and readings
- new feature: location-awareness (options -l -L -z)
- basic location-guessing based upon time zone
- coordinates may now also be in form degrees, minutes, seconds and be delimited by :, ', or "
- coordinates may now include compass directions N, S, E, W.
- time zone may now include fractional hours, or minutes (delimited by :, ', or ")
- highlight current day in reverse video
- compress width of display on screen
- print Gregorian and Hebrew months on one line
- option three-month to display previous and next months with current
- option three-month with year displays four rows, three columns
- option colorize to display data in, um, color
- option footnote to explain the month's holidays
- option force-hebrew for non-hebrew locales
- basic diaspora-awareness based on system time-zone

hdate
- introduced config file $XDG_CONFIG_HOME/hdate/hdaterc or ~/.config/hdate/hdaterc
- option --menu -m to prompt user for user-defined menu selections from config file
- option --candles and --havdalah take optional numeric arguments to modify the default minhag.
- remove dependency of Hebrew on locale environment variables
- new feature: sunset-aware to highlight correct hebrew date
- coordinates may now also be in form degrees, minutes, seconds and be delimited by :, ', or "
- coordinates may now include compass directions N, S, E, W.
- time zone may now include fractional hours, or minutes (delimited by :, ', or ")
- Hebrew information can optionally display in Hebrew characters in all locales
- option --sunset-aware, prints tomorrow's Hebrew date and info (for today only) if after sunset
- option --table, prints data in csv format, suitable for spreadsheets, awk, etc.

libhdate
- new function hdate_string
- DEPRECATION: hdate_get_day_string is now only a wrapper for hdate_string
- DEPRECATION: hdate_get_hebrew_month_string is now only a wrapper for hdate_string
- DEPRECATION: hdate_get_holyday_string is now only a wrapper for hdate_string
- DEPRECATION: hdate_get_month_string is now only a wrapper for hdate_string
- DEPRECATION: hdate_get_parasha_string is now only a wrapper for hdate_string
- DEPRECATION: hdate_get_omer_string is now only a wrapper for hdate_string
- DEPRECATION: hdate_get_int_string_ is now only a wrapper for hdate_string
- BUGFIX hdate_get_format_date, hdate-get_day_string, hdate_get_int_string no longer
  write to static local buffers. Callers to those functions must free().
- BUGFIX made all functions const-correct and fixed resulting conflicts

bindings - ruby
- BROKEN: We depend on 'swig', which is no longer supporting ruby 1.x.


3] Installation from source (Instructions)
==========================================

  This package is available for most major operating systems, so in
  most cases if you just want to use the command line programs, the
  following instructions are unnecessary; just install your
  distribution's pre-built package, or a packaged binary from a
  reliable repository compatable with your distribution.


3.1 Pre-Requisites
==================

  GNU toolchain, ie. gcc, libtoolize, autoreconf, and make. Your
  distribution may package these as `build-essential`, `libtool`,
  `autoconf`, and `automake`.

  This package was tested using the following toolchain versions:

    libtool (GNU libtool) 2.4.6
    autoconf (GNU Autoconf) 2.69
    automake (GNU automake) 1.16.1
    gcc (Debian 8.2.0-4) 8.2.0
    ldd (Debian GLIBC 2.27-5) 2.27

  Additonally, in order to build the bindings to other languages:
    swig 3.0.10
    g++ (Debian 8.2.0-4) 8.2.0
    g++-multilib 8.2.0  (for php)
    php 5.6.36
    python 2.7
    perl 5.26.2

3.2 Procedure
=============
  Run the provided shell script `autogen.sh`.

  The script will invoke `libtoolize`, `autoreconf`, `configure`, and
  `make`. Upon success, you can either use the built files at their
  current paths, or you can perform `sudo make install` to install
  them system-wide on your machine.

3.3 Extra credit
================

  The extensive manuals and documentation for the elements of the GNU
  toolchain are a reliable source of interminable travail,
  frustration, torture, and enlightenment for all those who would dare
  turn to them. The search engine of your choice also works.

ChangeLog
2022-01-16  Boruch Baum  <boruch_baum@gmx.com>

	* examples/hcal/hcal.c () Apply kludges to handle terminal emulators
		with bad bidi support. Add option --mlterm for strictly correct
		bidi suppport. Remove option --spacing as ugly. Convert many
		integer variables to boolean.

2022-01-12  Boruch Baum  <boruch_baum@gmx.com>

	* src/zdump3.c () Fix logic, especially for non-integer time zones.

	* examples/hcal/custom_days.c () set uniform symbol for Israeli
		national days.

2022-01-11  Boruch Baum  <boruch_baum@gmx.com>

	* examples/hcal/hdate.c () Correct quiet options for correct levels
		of message suppression. Remove compiler warnings. Set short form of
		Shabbat to zayin. Fix omer display. Update colophons and urls.
		Correct and update configure/make (thanks, Madhi!), fix time zone
		handling code.

2018-08-22  Boruch Baum  <boruch_baum@gmx.com>

	* USE: Update and correct usage example.

2018-08-27  Boruch Baum  <boruch_baum@gmx.com>

	* INSTALL: File removed. Content transferred to README

2018-08-22  Boruch Baum  <boruch_baum@gmx.com>

	* examples/hcal/hcal.c (print_footnote): bugfix print bidi
	footnote elements in correct sequence
	(html_print_day, print_calendar): initialize variables

	* examples/hcal/local_functions.c (set_default_location): Correct
	typo in longitude assignment


2018-08-19  Boruch Baum  <boruch_baum@gmx.com>

	* examples/hcal/local_functions.c (all): Convert tabs to
	two-space, remove doxygen? style three-slash comment forms, add
	comments, change whitespace
	(set_default_location): Correct timezone information for Honolulu,
	Sydney, Sao Paolo
	(get_epoch_time_range): Fix segfault, due to improper handling of
	null time zone string value.

	* examples/hcal/hcal.c, examples/hcal/hdate.c: Convert tabs to
	two-space, remove doxygen? style three-slash comment forms, add
	comments, change whitespace
	(print_week, print_dow_line): bugfix bidi print correct sequence
	(print_week): Fix segfault, due to improper handling of
	null time zone string value.
	(print_week): bugfix print shabbat name when using bidi

2018-08-17  Boruch Baum  <boruch_baum@gmx.com>

	* ALL: First commit since 2015, and using git!
	* ALL: Update copyrights and email-addresses
	* autogen.sh: New shel script file for building from sources
	* INSTALL: Replace generic ancient FSF file with useful, relevant one
	* USE: Edit for context
	* Makefile: Correct use of variables (CFLAGS->AM_CFLAGS, INCLUDE->AM_CPPFLAGS)
	* ALL: Remove unnecessary extra subdirectory level
	* Doxyfile, Doxyfile_cpp: Remove doxygen files
	* COPYING': Rename to 'LICENSE'
	* Changelog: Adopt a standard format

custom_days.c
- BUGFIX account for adjustments causing custom days to cross year boundaries

hdate.c
- BUGFIX custom_days respects -qqq option
- BUGFIX holidays display in hebrew and bidi
- BUGFIX custom days display in hebrew and bidi
- BUGFIX tabular holidays display in hebrew and bidi
- BUGFIX tabular custom days display in hebrew and bidi
- BUGFIX parasha display in hebrew and bidi
- BUGFIX parasha display data-first or label-first
- BUGFIX apostrophe for Shabbat prefix "yom sh"
- BUGFIX omer labels and english for -ooo
- BUGFIX omer -o display in hebrew and bidi
- BUGFIX all data labels convert spaces to underscores
- BUGFIX daf yomi masechtaot convert spaces to underscores
- BUGFIX omer convert spaces to underscores

hcal.c
- BUGFIX footnotes properly display in -gg mode

hdate_strings.c
- BUGFIX convert spaces to underscores (eg. אדר א, כי תשא)

he.po
- performed make update-po
- added and corrected many translations
----------------------------------------------------------------------------
hcal.c
- limits check for options -c -g --candles --havdalah

hdate.c
- limits check for options -q -t -o -v --candles --havdalah

libhdate.3
- continue transition from doxygen

----------------------------------------------------------------------------
hdate.c
- BUGFIX parse for user entry dd mm

hcal.c
- BUGFIX custom days print for html output

custom_days.c
- BUGFIX month IDs for Shabbat Chazon, Nachamu
- BUGFIX for custom day parsing and error reporting

BEGIN migration away from using doxygen

NOTE ON DOCUMENTATION STATE - libhdate seems unique in having
a man page for its header file, ie. apropos -s3 .h | grep "\.h (3)".
This will be removed, and replaced with a libhdate(3) man page

hdate_struct.3
- write this man page independent of doxygen, so that
  that it is comprehensible and useful.

libhdate.3
- new documentation file evolving from hdate.h.3
- added to version control
- update makefile.am

hdate.h.3
- deleted, removed from version control
- update makefile.am

----------------------------------------------------------------------------
hdate.c
- BUGFIX tabular month DST awareness
- Improve range checking for years, and julian days
- Improve range checking for base_years

hcal.c
- BUGFIX timezone alpha recognition
- BUGFIX html output
- Improve range checking for base_years

hdate.h
- tweak ranges for julian days

hcal.1
hdate.1
- updated

hdate_strings.c
- BUGFIX nlang_months indexing for short month names

----------------------------------------------------------------------------
hcal.c
hdate.c
- BUGFIX process_location_parms

hdate.c
- option epoch print in tabular mode
- BUGFIX sane output when user enters insane UTC adjustment
- BUGFIX sane output when user enters insane longitudes

local_functions.c
- BUGFIX process_location_parms
- improve alerts to user
- set default locations consistent with zone.tab file

hdate_strings.c
- more transliteration options for Hebrew months

----------------------------------------------------------------------------
hcal.c
- spelling error detected by lintian

hcal.1
hdate.1
- updated
- change hyphens to minus signs (lintian-compliance)

hdate.h.3
- tweak so lintian doesn't reject SH macro
- change hyphens to minus signs (lintian-compliance)

----------------------------------------------------------------------------
custom_days.c
- improved file handling

src/hdate_holyday.c
/*  libhdate - Hebrew calendar library: http://libhdate.sourceforge.net
 *
 *  Copyright (C) 2011-2018 Boruch Baum  <boruch_baum@gmx.com>
 *                2004-2007 Yaacov Zamir <kzamir@walla.co.il>
 *                1984-2003 Amos Shapir
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "hdate.h"
#include "support.h"

/**
 @brief Return number of halachic holiday or Israeli custom day.

 @param h The hdate_struct of the date to use.
 @param diaspora if True give diaspora holydays
 @return the number of holyday.
*/
int
hdate_get_holyday (hdate_struct const * h, int diaspora)
{
	int day_code = 0;
	day_code = hdate_get_halachic_day( h, diaspora);
	if (day_code == 0)
		day_code = hdate_get_israeli_day( h, diaspora);
	return day_code;
}


/**
 @brief Return number of hebrew halachic holiday.

 @param h The hdate_struct of the date to use.
 @param diaspora if True give diaspora holydays
 @return the number of holyday.
*/
int
hdate_get_halachic_day (hdate_struct const * h, int diaspora)
{
	int halachic_day;

	/* halachic_days table */
	static int halachic_days_table[14][30] =
	{
		{	/* Tishrey */
			1, 2, 3, 3, 0, 0, 0, 0, 37, 4,
			0, 0, 0, 39, 5, 31, 6, 6, 6, 6,
			7, 27, 8, 0, 0, 0, 0, 0, 0, 0},
		{	/* Heshvan */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Kislev */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 9, 9, 9, 9, 9, 9},
		{	/* Tevet */
			9, 9, 9, 0, 0, 0, 0, 0, 0, 10,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Shvat */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 11, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Adar */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			12, 0, 12, 13, 14, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Nisan */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 38, 15, 32, 16, 16, 16, 16,
			28, 29, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Iyar */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 18, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Sivan */
			0, 0, 0, 0, 19, 20, 30, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Tamuz */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 21, 21, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Av */
			0, 0, 0, 0, 0, 0, 0, 0, 22, 22,
			0, 0, 0, 0, 23, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Elul */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Adar 1 */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Adar 2 */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			12, 0, 12, 13, 14, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
	};

	/* sanity check */
	if (h->hd_mon < 1 || h->hd_mon > 14 || h->hd_day < 1 || h->hd_day > 30)
		return 0;

	halachic_day = halachic_days_table[h->hd_mon - 1][h->hd_day - 1];

	/* if tzom on sat delay one day */
	/* tzom gdalyaho on sat */
	if ((halachic_day == 3) && (h->hd_dw == 7 || (h->hd_day == 4 && h->hd_dw !=1)))
		halachic_day = 0;
	/* 17 of Tamuz on sat */
	if ((halachic_day == 21) && ((h->hd_dw == 7) || (h->hd_day == 18 && h->hd_dw != 1)))
		halachic_day = 0;
	/* 9 of Av on sat */
	if ((halachic_day == 22) && ((h->hd_dw == 7) || (h->hd_day == 10 && h->hd_dw != 1)))
		halachic_day = 0;

	/* Hanukah in a long year */
	if ((halachic_day == 9) && (h->hd_size_of_year % 10 != 3) && (h->hd_day == 3))
		halachic_day = 0;

	/* if tanit ester on sat mov to Thu */
	if ((halachic_day == 12) && ((h->hd_dw == 7) || (h->hd_day == 11 && h->hd_dw != 5)))
		halachic_day = 0;

	/* diaspora holidays */

	/* simchat tora only in diaspora in israel just one day shmini+simchat tora */
	if (halachic_day == 8 && !diaspora)
		halachic_day = 0;

	/* sukkot II holiday only in diaspora */
        if (halachic_day == 31 && !diaspora)
                halachic_day = 6;

	/* pesach II holiday only in diaspora */
        if (halachic_day == 32 && !diaspora)
                halachic_day = 16;

	/* shavot II holiday only in diaspora */
	if (halachic_day == 30 && !diaspora)
		halachic_day = 0;

	/* pesach VIII holiday only in diaspora */
	if (halachic_day == 29 && !diaspora)
		halachic_day = 0;

	return halachic_day;
}

/**
 @brief Return number of Israeli custom day.

 @param h The hdate_struct of the date to use.
 @param diaspora if True give diaspora holydays
 @return the number of holyday.
*/
int
hdate_get_israeli_day (hdate_struct const * h, int diaspora)
{
	int israeli_day;

	/* holydays table */
	static int israeli_days_table[14][30] =
	{
		{	/* Tishrey */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Heshvan */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 35,
			35, 35, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Kislev */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Tevet */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Shvat */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 33},
		{	/* Adar */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Nisan */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 24, 24, 24, 0, 0},
		{	/* Iyar */
			0, 17, 17, 17, 17, 17, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 26, 0, 0},
		{	/* Sivan */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Tamuz */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 36, 36},
		{	/* Av */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Elul */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Adar 1 */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
		{	/* Adar 2 */
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
	};

	/* sanity check */
	if (h->hd_mon < 1 || h->hd_mon > 14 || h->hd_day < 1 || h->hd_day > 30)
		return 0;

	israeli_day = israeli_days_table[h->hd_mon - 1][h->hd_day - 1];

	/* yom yerushalym after 68 */
	if (israeli_day == 26)
	{
		if (h->gd_year < 1968)
			israeli_day = 0;
	}

	/* yom ha azmaot and yom ha zicaron */
	if (israeli_day == 17)
	{
		if (h->gd_year < 1948)
			israeli_day = 0;
		else if (h->gd_year < 2004)
		{
			if ((h->hd_day == 3) && (h->hd_dw == 5))
				israeli_day = 17;
			else if ((h->hd_day == 4) && (h->hd_dw == 5))
				israeli_day = 17;
			else if ((h->hd_day == 5) && (h->hd_dw != 6 && h->hd_dw != 7))
				israeli_day = 17;
			else if ((h->hd_day == 2) && (h->hd_dw == 4))
				israeli_day = 25;
			else if ((h->hd_day == 3) && (h->hd_dw == 4))
				israeli_day = 25;
			else if ((h->hd_day == 4) && (h->hd_dw != 5 && h->hd_dw != 6))
				israeli_day = 25;
			else
				israeli_day = 0;
		}
		else
		{
			if ((h->hd_day == 3) && (h->hd_dw == 5))
				israeli_day = 17;
			else if ((h->hd_day == 4) && (h->hd_dw == 5))
				israeli_day = 17;
			else if ((h->hd_day == 6) && (h->hd_dw == 3))
				israeli_day = 17;
			else if ((h->hd_day == 5) && (h->hd_dw != 6 && h->hd_dw != 7 && h->hd_dw != 2))
				israeli_day = 17;
			else if ((h->hd_day == 2) && (h->hd_dw == 4))
				israeli_day = 25;
			else if ((h->hd_day == 3) && (h->hd_dw == 4))
				israeli_day = 25;
			else if ((h->hd_day == 5) && (h->hd_dw == 2))
				israeli_day = 25;
			else if ((h->hd_day == 4) && (h->hd_dw != 5 && h->hd_dw != 6 && h->hd_dw != 1))
				israeli_day = 25;
			else
				israeli_day = 0;
		}
	}

	/* yom ha shoaa, on years after 1958 */
	if (israeli_day == 24)
	{
		if (h->gd_year < 1958)
			israeli_day = 0;
		else
		{
			// if 27 Nissan is Fri, commemorate on prior Thursday
			// if 27 Nissan is Sun, commemorate on following Monday
			if ((h->hd_day == 26) && (h->hd_dw != 5))
				israeli_day = 0;
			if ((h->hd_day == 28) && (h->hd_dw != 2))
				israeli_day = 0;
			if ((h->hd_day == 27) && (h->hd_dw == 6 || h->hd_dw == 1))
				israeli_day = 0;
		}
	}

	/* Rabin day, on years after 1997 */
	if (israeli_day == 35)
	{
		if (h->gd_year < 1997)
			israeli_day = 0;
		else
		{
			// if 12 Cheshvan is Fri/Sat, commemorate on prior Thursday
			if ((h->hd_day == 10 || h->hd_day == 11) && (h->hd_dw != 5))
				israeli_day = 0;
			if ((h->hd_day == 12) && (h->hd_dw == 6 || h->hd_dw == 7))
				israeli_day = 0;
		}
	}

	/* Zhabotinsky day, on years after 2005 */
	if (israeli_day == 36)
	{
		if (h->gd_year < 2005)
			israeli_day = 0;
		else
		{
			// if 29 Tammuz is Shabbat, commemorate on following Sunday
			if ((h->hd_day == 30) && (h->hd_dw != 1))
				israeli_day = 0;
			if ((h->hd_day == 29) && (h->hd_dw == 7))
				israeli_day = 0;
		}
	}

	return israeli_day;
}


/**
 @brief Return the day in the omer of the given date

 @param h The hdate_struct of the date to use.
 @return The day in the omer, starting from 1 (or 0 if not in sfirat ha omer)
*/
int
hdate_get_omer_day(hdate_struct const * h)
{
	int omer_day;
	hdate_struct sixteen_nissan;

	hdate_set_hdate(&sixteen_nissan, 16, 7, h->hd_year);
	omer_day = h->hd_jd - sixteen_nissan.hd_jd + 1;

	if ((omer_day > 49) || (omer_day < 0))
		omer_day = 0;

	return omer_day;
}

/**
 @brief Return number of hebrew holyday type.

  Holiday types:
    0 - Regular day
    1 - Yom tov (plus yom kippor)
    2 - Erev yom kippur
    3 - Hol hamoed
    4 - Hanuka and purim
    5 - Tzomot
    6 - Independance day and Yom yerushalaim
    7 - Lag baomer ,Tu beav, Tu beshvat
    8 - Tzahal and Holocaust memorial days
    9 - National days

 @param holyday the holyday number
 @return the number of holyday type.
*/

int
hdate_get_holyday_type (int day_code)
{
	int day_type = 0;
	day_type = hdate_get_halachic_day_type ( day_code );
	if (day_type == 0)
		day_type = hdate_get_israeli_day_type ( day_code );
	return day_type;
}


/**
 @brief Return number of hebrew halachic day type.

  Holiday types:
    0 - Regular day
    1 - Yom tov (plus yom kippor)
    2 - Erev yom kippur
    3 - Hol hamoed
    4 - Hanuka and purim
    5 - Tzomot
    6 - Lag baomer ,Tu beav, Tu beshvat

 @param halachic_day the holyday number
 @return the number of halachic_day type.
*/

int
hdate_get_halachic_day_type (int halachic_day)
{
	int halachic_day_type;

	switch (halachic_day)
	{
	case 0: /* regular day */
		halachic_day_type = 0;
		break;

	// Type 1 - Yom Tov
	case 1: // Rosh Hashana (first day)
	case 2: // Rosh Hashana (second day)
	case 4: // Yom Kippur
	case 5: // Sukkot
	case 8: // Simchat Torah
	case 15:// Pesach
	case 20:// Shavuot
	case 27:// Shmini Atzeret
	case 28:// Shevi'i shel Pesach
	case 29:// Acharon shel Pesach
	case 30:// Shavuot (second day)
	case 31:// Sukkot (second day)
	case 32:// Pesach (second day)
	// An old comment said: "To find erev yom tov, check if tomorrow returns 1";
	// However, I didn't see that being done anywhere - Boruch
		halachic_day_type = 1;
		break;

	// Type 2 - Erev Yom Tov
	case 19: /* Erev Shavuot */
	case 37: /* Erev Yom Kippur */
	case 38: /* Erev Pesach */
	case 39: /* Erev Sukkot */
		halachic_day_type = 2;
		break;

	// Type 3 - Chol HaMoed
	case 6: // Hol hamoed Sukkot
	case 7: // Hoshana raba
	case 16:// Hol hamoed Pesach
		halachic_day_type = 3;
		break;

	// Type 4 - Chanukah and Purim
	case 9: // Chanukah
	case 13:// Purim
	case 14:// Shushan Purim
		halachic_day_type = 4;
		break;

	// Type 5 - Fast days
	case 3: // Tzom Gedaliah
	case 10:// Asara B'Tevet
	case 12:// Ta'anit Esther
	case 21:// Tzom Tammuz
	case 22:// Tish'a B'Av
		halachic_day_type = 5;
		break;

	// Type 6 -
	case 18:// Lag B'Omer
	case 23:// Tu B'Av
	case 11:// Tu B'Shvat
		halachic_day_type = 6;
		break;
	}

	return halachic_day_type;
}

/**
 @brief Return number of israeli_day type.

  Holiday types:
    0 - Regular day
    6 - Independance day and Yom yerushalaim
    8 - Tzahal and Holocaust memorial days
    9 - National days

 @param israeli_day the israeli_day number
 @return the number of israeli_day type.
*/

int
hdate_get_israeli_day_type (int israeli_day)
{
	int israeli_day_type;

	switch (israeli_day)
	{
	case 0: /* regular day */
		israeli_day_type = 0;
		break;

	// Type 6 -
	case 17:// Yom HaAtzma'ut
	case 26:// Yom Yerushalayim
		israeli_day_type = 6;
		break;

	// Type 8 -
	case 24:// Yom HaShoah
	case 25:// Yom HaZikaron
		israeli_day_type = 8;
		break;


	// Type 9 - Default
	case 33:// Family Day (30 Shvat)
	case 34:// Memorial day for fallen whose place of burial is unknown
			// BUGFIX - The date also seems to be unknown
	case 35:// Yitzhak Rabin memorial day (mid-Cheshvan)
	case 36:// Zeev Zhabotinsky day (end-Tammuz)
	default: /* National days */
		israeli_day_type = 9;
		break;
	}

	return israeli_day_type;
}
